### ISSUE TITLE
Greedy sampling disagrees with its oracle on the row-0 tie

### ISSUE THREAD

**Issue by sebastianrcnt** [sebastianrcnt/ninfer-3090#4]
## Symptom

`ninfer_turboquant_gqa_test`'s sibling `ninfer_sampling_test` fails deterministically on `main`
(`88b249c`), RTX 3090 / CUDA 13.1:

```
sample greedy mathematical result: exact mismatch at index 0
    sample top-k/top-p/min-p FP64 distribution match (max z=1.42365)
    sample top-k public cap FP64 distribution match (max z=1.77583)
    sample real token-domain B=8 FP64 distribution match (max z=0.863831)
FAIL sample public contract
```

Only `greedy_contract`'s first assertion fails. Every distribution check passes, the
`sample greedy skips token_counts updates` check passes, and the failure reproduces on three
consecutive runs — this is not sampling noise.

`index 0` is the batch row, not a token id: row 0 of 8 picks a different token than the oracle.

## Why row 0 is the interesting one

`greedy_contract` (`tests/ops/test_sampling.cpp:321`) builds each row with two elevated logits
and leaves the rest at `-9.0f`:

```cpp
int first  = (17 + 7919 * row) % token_domain;
int second = token_domain - 1 - ((31 + 65537 * row) % token_domain);
...
logits[base + first]  = row == 0 ? -0.0f : 16.0f + row;
logits[base + second] = row == 0 ?  0.0f : 16.0f + row;
```

Row 0 is the only row where the two candidates are **not** separated by value. It is also the
only row that is failing. For `row == 0`:

* `first  = 17`
* `second = 248,045`
* `logits[17] = -0.0f`, `logits[248045] = +0.0f`

Both the oracle and the device comparators break ties toward the lower index, so both should
choose 17:

```cpp
// tests/ops/test_sampling.cpp:69 — strict >, so an equal value never displaces an earlier index
if (logits[base + token] > logits[base + best]) { best = token; }

// src/ops/kernel/argmax.cuh:24 and src/ops/kernel/sampling_device.cuh:77
return value > best_value || (value == best_value && index < best_index);
```

## Two candidate causes

They predict the same wrong answer (248,045), so the failure alone does not separate them.

**1. Signed zero.** `-0.0f == +0.0f` is true in IEEE and `+0.0f > -0.0f` is false, so a
value-based comparison keeps index 17. Any path that compares BF16 *bit patterns* instead — a
packed key, an integer `max`, a sort on raw bits — orders `+0.0` (`0x0000`) above `-0.0`
(`0x8000`) and selects 248,045.

**2. Penalties applied under greedy.** The config sets `presence_penalty = 100.0f`,
`frequency_penalty = 100.0f` with `counts[17] = 9`, and `first` for row 0 **is token 17** — the
single token carrying a nonzero count. The oracle ignores penalties entirely, so the contract is
that greedy returns the raw argmax. If the greedy path subtracts `100 + 9 * 100` from token 17
first, it drops far below `+0.0` and 248,045 wins.

The coincidence that row 0's tie is between the one penalized token and an untouched one looks
deliberate — the case appears designed to catch exactly this.

## Discriminating experiment

One run each, no fix required:

* clear `counts[17]` (or zero both penalties) and rerun. Still failing ⇒ cause 1; passing ⇒ cause 2.
* independently, set row 0's pair to two equal nonzero values, e.g. `1.0f`/`1.0f`. Still failing
  ⇒ tie-breaking is wrong in general, not just at signed zero.

## Notes

Pre-existing and unrelated to recent work: it fails identically with `fix/turboquant-decode-unclamped-window`
stashed, and at `4676573`. Surfaced while establishing a test baseline for #1.
